src/sparsebase/utils/converter/converter.h

```diff
@@ -43,124 +45,279 @@ namespace converter {
+                  std::type_index to_type, const std::vector<context::Context *>&to_contexts,
+                  bool is_move_conversion = false);
+  //! Clears all conversion functions from one format to another.
   /*! Removes conversion functions that convert from `from_type` to `to_type` from
    * the conversion map
    * @param from_type type_index of the original format type
    * @param to_type type_index of the destination format type
    * @param move_conversion whether the conversions to be removed are move conversions or not
    */
   void ClearConversionFunctions(std::type_index from_type, std::type_index to_type, bool move_conversion = false);
-  
-  /*! Removes all conversion functions from the current converter
+
+  //! Removes all conversion functions from current converter.
+  /*! Removes all conversion functions from the current converter in all directions.
    * @param move_conversion whether the conversions to be removed are move conversions or not
    */
   void ClearConversionFunctions(bool move_conversion = false);
-  
-  std::vector<format::Format *>
-  ApplyConversionSchema(ConversionSchemaConditional cs,
-                        std::vector<format::Format *> packed_sfs,
+
+  //! Applies a conversion schema to convert one format to another.
+  /*!
+   * Takes a conversion schema to convert all the formats in `packed_sfs`, carries out the conversions,
+   * and returns a chain of converted formats for each format. Chain contains original format, all intermediate formats,
+   * and last format in the chain.
+   * @param cs ConversionSchema containing the an optional chain of conversion functions for every format in `packed_sfs`
+   * @param packed_sfs formats to convert
+   * @param is_move_conversion whether conversion is move conversion
+   * @return a vector of size `packed_sfs.size()`. Each element is a vector of size >= 1 containing the starting format
+   * in the conversion chain, every intermediate format, and the final format. The first format is the original and the last
+   * is the output format in each chain.
+   */
+  static std::vector<std::vector<format::Format *>>
+  ApplyConversionSchema(const ConversionSchema & cs,
+                        const std::vector<format::Format *> &packed_sfs,
                         bool is_move_conversion = false);
+
+   //! Takes a conversion chain and a format and applies that chain on the format
+   //! to produce some output format.
+   /*!
+    * \param chain a conversion chain containing conversion functions and contexts to use for each function.
+    * \param is_move_conversion whether it's a move conversion or not.
+    * \return a vector of format with the first being the original format, the last being the target format,
+    * and the rest being intermediate formats. If a conversion is empty or false, only returns the original format.
+    */
+  static std::vector<format::Format*> ApplyConversionChain(const ConversionChain& chain,
```

Comment:
So if move conversion is enabled, only the last format of this vector is valid. But if it's not, doesn't it contain duplicates?

Reply:
It does. Technically, all the formats in the chain will be different representations of the same logical Format object. However, if users don't want these formats, they will just get deleted at the next step in the execution (`Execute`). 

I see now that users might want to save memory by deleting a format as soon as it is not needed. Maybe we can replace the move options with such an option? I mean, if someone wants to move convert, they probably don't want the intermediate formats, right? What do you think?

Reply:
Yeah that was my though process as well

Reply:
Makes sense. Let me add that.

Reply:
Okay, so here's what I did:
* For `ApplyConversionChain`, I replaced `is_move_conversion` with `clear_intermediate`. 
  * `is_move_conversion` wasn't being used to begin with since at that point, the conversion functions were already decided by `GetConversionChain`.
  * `clear_intermediate` would delete intermediate formats after each step, keeping only the start format and destination format.
* Replaced `is_move_conversion` with `clear_intermediate` in `ApplyConversionSchema` for the same reason as above.
* In `FunctionMatcher`, added a `clear_conversion` parameter to `CachedExecute`. Here is what happens when it's true and when it's false:
  * `clear_intermediate == true`: if a chain size > 1, all the formats between the first and the last are deleted. It will only return the destination format. (In contrast, Execute doesn't return _any_ formats)
  * `clear_intermediate == false`: will return all the destination format and all the intermediate formats.
* In `FunctionMatcher` in `Execute`, the call to `CachedExecute` will always clear intermediate formats. Then inside `Execute`, the destination formats are also deleted.

Now, I want to retract my statement that "if someone wants to move convert they probably don't want the intermediate formats." That can actually be the case. Imagine move converting CSC->CSR. This will do CSC -> COO -> CSR. However, if move conversions are used, we will end up with three objects but only 6 arrays:
`row_ptr` (CSR and COO share this)
`col_ptr` (CSC)
`col` (CSR and COO share this)
`vals` (COO and CSR share this)
`row` (CSC)
`vals` (CSR and COO share this)
`vals_csc` (CSC) 
(CSC vals and the other two vals are different because of how edges are sorted)

Which is what the "view" idea would've required. So I think clearing intermediate and moving should be separate ideas.

What do you think?

Reply:
Ok looks good to me